Consecutive call of nvim_get_mode() does not return the same mode

### Neovim Version

NVIM v0.6.0-dev

### Operating system/version:

Ubuntu 20.04.2

### Terminal name/version:

Terminator 1.91

### TERM environment variable

xterm-256color

### Installation

build from repo

### Steps to reproduce

Run the following code

```python
import pynvim

n = pynvim.attach(
    "child", argv=['/bin/env', 'nvim', '--clean', '--embed', '--headless', 'filename'])

n.input(":set cmdheight=2<CR>")

n.feedkeys(n.replace_termcodes("gh"), 't')
assert (n.api.get_mode() == dict(
    {'mode': 's', 'blocking': False})), "not in select mode or is blocked!"
assert (n.api.get_mode() == dict(
    {'mode': 's', 'blocking': False})), "not in select mode or is blocked!"

n.feedkeys(n.replace_termcodes("<C-o>"), 't')
assert (n.api.get_mode() == dict(
    {'mode': 'vs', 'blocking': False})), "not in select visual mode or is blocked!"
assert (n.api.get_mode() == dict(
    {'mode': 'vs', 'blocking': False})), "not in select visual mode or is blocked!"

n.feedkeys(n.replace_termcodes("<C-o>"), 'xt')
assert (n.api.get_mode() == dict(
    {'mode': 's', 'blocking': False})), "not in select mode or is blocked!"
assert (n.api.get_mode() == dict(
    {'mode': 's', 'blocking': False})), "not in select mode or is blocked!"

n.feedkeys(n.replace_termcodes("<C-o>"), 't')
assert (n.api.get_mode() == dict(
    {'mode': 'vs', 'blocking': False})), "not in select visual mode or is blocked!"
assert (n.api.get_mode() == dict(
    {'mode': 'vs', 'blocking': False})), "not in select visual mode or is blocked!"

n.close()
```

```
Traceback (most recent call last):
  File "test/test-mode.py", line 18, in <module>
    assert (n.api.get_mode() == dict(
AssertionError: not in select visual mode or is blocked!
```

### Vim

N/A

### Expected behavior

Return the same mode on consecutive calls

### Actual behavior

Returns select mode from visual select mode when doing nothing

--- a/test/functional/editor/mode_visual_spec.lua
+++ b/test/functional/editor/mode_visual_spec.lua
@@ -14,6 +14,7 @@ describe('visual-mode', function()
   it("select-mode Ctrl-O doesn't cancel Ctrl-O mode when processing event #15688", function()
     feed('iHello World<esc>gh<c-o>')
     eq({mode='vs', blocking=false}, meths.get_mode()) -- fast event
+    eq({mode='vs', blocking=false}, meths.get_mode()) -- again #15288
     eq(2, eval('1+1'))  -- causes K_EVENT key
     eq({mode='vs', blocking=false}, meths.get_mode()) -- still in ctrl-o mode
     feed('^')

